mlflow/store/db_artifact_repo.py

```diff
@@ -0,0 +1,279 @@
+        """
+        Return all the artifacts for this run_id directly under path. If path is a file, returns
+        an empty list. Will error if path is neither a file nor directory.
+
+        :param path: Relative source path that contains desired artifacts
+
+        :return: List of artifacts as FileInfo listed directly under path.
+        """
+        path = os.path.normpath(path)
+        with self.ManagedSessionMaker() as session:
+            regex = path + os.sep + '%'
+            return [artifact.to_file_info()
+                    for artifact in
+                    session.query(SqlArtifact).filter(
+                        or_(SqlArtifact.group_path.like(os.path.join(self.root, regex)),
+                            SqlArtifact.group_path == path))]
+
+    def download_artifacts(self, artifact_path, dst_path=None):
+        """
+        Download an artifact file or directory to a local directory if applicable, and return a
+        local path for it.
+        The caller is responsible for managing the lifecycle of the downloaded artifacts.
+
+        :param artifact_path: Relative source path to the desired artifacts.
+        :param dst_path: Absolute path of the local filesystem destination directory to which to
+                         download the specified artifacts. This directory must already exist.
+                         If unspecified, the artifacts will either be downloaded to a new
+                         uniquely-named directory on the local filesystem or will be returned
+                         directly in the case of the LocalArtifactRepository.
+
+        :return: Absolute path of the local filesystem location containing the desired artifacts.
+        """
+
+        # TODO: Probably need to add a more efficient method to stream just a single artifact
+        #       without downloading it, or to get a pre-signed URL for cloud storage.
+
+        def download_artifacts_into(artifact_path, dest_dir):
+            basename = os.path.basename(artifact_path)
+            local_path = os.path.join(dest_dir, basename)
+
+            listing = self.list_artifacts(artifact_path)
+            if len(listing) > 0:
+                # Artifact_path is a directory, so make a directory for it and download everything
+                if not os.path.exists(local_path):
+                    os.mkdir(local_path)
+                for file_info in listing:
+                    # prevent an infinite loop (sometimes the current path is listed e.g. as ".")
+                    if file_info.path == "." or file_info.path == artifact_path:
+                        continue
+                    download_artifacts_into(artifact_path=file_info.path, dest_dir=local_path)
+            else:
+                self._download_file(remote_file_path=artifact_path, local_path=local_path)
+            return local_path
+
+        if artifact_path:
+            artifact_path = os.path.normpath(artifact_path)
```

Comment:
Is there a particular reason for using `normpath` here rather than `abspath`? Otherwise, the entire `download_artifacts` method seems identical to `ArtifactRepository.download_artifacts` and can probably be removed, provided `normpath` is not necessary.

Reply:
@avflor Can you provide input here?

Reply:
The `ArtifactRepository.download_artifacts` does not apply `normpath `or `abspath `to the input `artifact_path` but only to `dst_path `. I've tried using the default one but I get test failures on windows. Example error:

```
> self = <tests.store.test_db_artifact_repo.TestSqlAlchemyStoreSqlite testMethod=test_download_file_artifact>
> 
>     def test_download_file_artifact(self):
> 
>         with TempDir() as root_dir:
> 
>             with open(root_dir.path("file_one.txt"), "w") as f:
> 
>                 f.write('DB store Test One')
> 
>     
> 
>             self.store.log_artifacts(root_dir._path, 'new_path/path')
> 
>             local_path = self.store.download_artifacts(artifact_path='new_path/path/file_one.txt')
> 
> >           assert open(local_path).read() == 'DB store Test One'
> 
> E           FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\travis\\AppData\\Local\\Temp\\tmpb8oe5tyk\\file_one.txt'
> 
> tests\store\test_db_artifact_repo.py:186: FileNotFoundError
```

> 
> 
> @avflor Can you provide input here?